### ISSUE TITLE
Request: warn when converting string to byte slice when appending to byte slice

### ISSUE THREAD

**Issue by mark-rushakoff** [golang/lint#264]
I have come across some code that was appending a string literal or a string value to a byte slice, but converted the string to a byte slice inside the call to `append`. 

As far as I'm aware, nothing is gained by converting these strings to byte slices for the sake of `append`.  I think this pattern is deserving of a lint warning. I'm not sure how common it is in the wild, but I believe if someone writes this intentionally, they may not be aware that appending the string directly is supported. I believe this warning does not belong in vet because behavior should be functionally the same regardless of whether appending a byte slice or a string.

Here are some examples of code that I believe should warn ([Playground link](https://play.golang.org/p/n0xf4Y8Cxl)):

```go
var buf []byte

// Unnecessary conversion, string literal.
buf = append(buf, []byte("hello")...)

// Unnecessary conversion, string constant.
const S = "hello"
buf = append(buf, []byte(S)...)

// Unnecessary conversion, result of function call returning string.
stringer := func() string { return "hello" }
buf = append(buf, []byte(stringer())...)

// Unnecessary conversion, constant whose type's underlying type is string.
type MyString string
const ms MyString = "hello"
buf = append(buf, []byte(ms)...)
```

Although perhaps not strictly relevant for lint, here is a short benchmark that demonstrates that converting the string to a byte slice in a call to `append` is slower than not converting the string:

```go
package bench_test

import "testing"

var sink interface{}

func BenchmarkAppendStringLiteral(b *testing.B) {
	buf := make([]byte, 0, 10)
	for i := 0; i < b.N; i++ {
		buf = append(buf, "hello"...)
		sink = buf
		buf = buf[:0]
	}
}

func BenchmarkAppendStringByteSlice(b *testing.B) {
	buf := make([]byte, 0, 10)
	for i := 0; i < b.N; i++ {
		// The following line should warn about an unnecessary conversion.
		buf = append(buf, []byte("hello")...)
		sink = buf
		buf = buf[:0]
	}
}
```

```
# $GOPATH/bin/go1.8beta2 env
GOARCH="amd64"
GOBIN=""
GOEXE=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOOS="darwin"
GOPATH="/Users/mr/go"
GORACE=""
GOROOT="/Users/mr/sdk/go1.8beta2"
GOTOOLDIR="/Users/mr/sdk/go1.8beta2/pkg/tool/darwin_amd64"
GCCGO="gccgo"
CC="clang"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/ct/bl4_z3g51ks8239_r2k07v_40000gn/T/go-build292155367=/tmp/go-build -gno-record-gcc-switches -fno-common"
CXX="clang++"
CGO_ENABLED="1"
PKG_CONFIG="pkg-config"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"

# $GOPATH/bin/go1.8beta2 test -bench=. -benchtime=5s -benchmem
BenchmarkAppendStringLiteral-4     	200000000	        47.0 ns/op	      32 B/op	       1 allocs/op
BenchmarkAppendStringByteSlice-4   	100000000	        57.3 ns/op	      32 B/op	       1 allocs/op
PASS

# go env
GOARCH="amd64"
GOBIN=""
GOEXE=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOOS="darwin"
GOPATH="/Users/mr/go"
GORACE=""
GOROOT="/usr/local/Cellar/go/1.7.3/libexec"
GOTOOLDIR="/usr/local/Cellar/go/1.7.3/libexec/pkg/tool/darwin_amd64"
CC="clang"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/ct/bl4_z3g51ks8239_r2k07v_40000gn/T/go-build759449954=/tmp/go-build -gno-record-gcc-switches -fno-common"
CXX="clang++"
CGO_ENABLED="1"

# go test -bench=. -benchtime=5s -benchmem
testing: warning: no tests to run
BenchmarkAppendStringLiteral-4     	200000000	        47.7 ns/op	      32 B/op	       1 allocs/op
BenchmarkAppendStringByteSlice-4   	100000000	        57.1 ns/op	      32 B/op	       1 allocs/op
PASS
```

**Comment by dsnet** [comment]
This check seems pretty reasonable to me. I should also mention that `copy` also has the special case where `strings` can be copied into `[]byte`.

**Comment by dominikh** [comment]
@dsnet Do you want this in golint or in gosimple?

**Comment by dsnet** [comment]
I think this is reasonable to include in golint since it is stylistically nicer to read `append(b, s...)` than it is to read `append(b, []byte(s)...)`.
